File: internal/capability/audit_test.go
package capability

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/ro-ag/ptrack/internal/model"
	"github.com/ro-ag/ptrack/internal/store"
)

func TestAuditRecordIsBoundedAndRedactedInDatabase(t *testing.T) {
	dbPath := filepath.Join(t.TempDir(), "ptrack.db")
	s, err := store.Open(dbPath)
	if err != nil {
		t.Fatal(err)
	}
	capability := model.Capability{
		ID: 42, Kind: model.CapabilityHTTP, AgentProfile: "agent-codex",
		Audit: model.CapabilityAuditPolicy{Enabled: true, RetainLast: 2},
	}
	recorder := Recorder{Store: s, Now: func() time.Time { return time.Unix(1, 0) }}
	secret := "CANARY_SUPER_SECRET_TOKEN"
	for index := 0; index < 3; index++ {
		err := recorder.Record(context.Background(), capability, AuditEvent{
			Operation:     "GET",
			Target:        "https://user:" + secret + "@example.com/api/" + secret + "?token=" + secret,
			ErrorClass:    secret,
			Duration:      48 * time.Hour,
			RequestBytes:  -1,
			ResponseBytes: 1 << 50,
			Redirects:     100,
		})
		if err != nil {
			t.Fatal(err)
		}
	}
	audits, err := s.ListCapabilityAudits(capability.ID, 0)
	if err != nil {
		t.Fatal(err)
	}
	if len(audits) != 2 {
		t.Fatalf("audit count = %d want 2", len(audits))
	}
	got := audits[0]
	if got.Target != "https://example.com" || strings.Contains(got.Target, secret) || got.ErrorClass != "internal" || got.DurationMillis != (24*time.Hour).Milliseconds() || got.RequestBytes != 0 || got.ResponseBytes != 1<<40 || got.Redirects != maxRedirects {
		t.Fatalf("audit was not sanitized: %+v", got)
	}
	if err := s.Close(); err != nil {
		t.Fatal(err)
	}
	raw, err := os.ReadFile(dbPath)
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(string(raw), secret) {
		t.Fatal("secret-bearing audit input persisted in raw database")
	}
}

func TestDisabledAuditDoesNotWrite(t *testing.T) {
	s, err := store.Open(filepath.Join(t.TempDir(), "ptrack.db"))
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()
	recorder := Recorder{Store: s}
	if err := recorder.Record(context.Background(), model.Capability{ID: 9}, AuditEvent{Operation: "GET"}); err != nil {
		t.Fatal(err)
	}
	audits, err := s.ListCapabilityAudits(9, 0)
	if err != nil || len(audits) != 0 {
		t.Fatalf("disabled audit wrote records: %+v, %v", audits, err)
	}
}

File: crates/ptrack-capability/src/audit_test.rs
use ptrack_capability_policy::AuditEvent;

use super::audit::AuditRecorder;
use super::test_support::{TempDir, approved_http, store};

#[test]
fn recorder_skips_absent_store_and_disabled_policy() {
    let capability = approved_http("https://example.com/api");
    AuditRecorder::new(None)
        .record(&capability, &event(true))
        .unwrap();

    let temp = TempDir::new("audit-disabled");
    let store = store(&temp);
    let mut disabled = capability;
    disabled.audit.enabled = false;
    AuditRecorder::new(Some(&store))
        .record(&disabled, &event(true))
        .unwrap();
    assert!(store.capability_audits(0, 0).unwrap().is_empty());
}

#[test]
fn recorder_persists_only_sanitized_bounded_shape() {
    let temp = TempDir::new("audit-sanitized");
    let store = store(&temp);
    let capability = approved_http("https://example.com/api");
    AuditRecorder::new(Some(&store))
        .record(&capability, &event(false))
        .unwrap();
    let audits = store.capability_audits(capability.id, 1).unwrap();
    let audit = &audits[0];
    assert_eq!(audit.capability_id, capability.id);
    assert_eq!(audit.agent_profile, "agent-codex");
    assert_eq!(audit.operation, "unknown");
    assert_eq!(audit.target, "https://example.com");
    assert_eq!(audit.error_class, "internal");
    assert_eq!(audit.duration_millis, 86_400_000);
    assert_eq!(audit.request_bytes, 0);
    assert_eq!(audit.response_bytes, 1_i64 << 40);
    assert_eq!(audit.redirects, 10);
}

#[test]
fn recorder_reports_sanitized_storage_failure() {
    let temp = TempDir::new("audit-error");
    let store = store(&temp);
    let mut capability = approved_http("https://example.com/api");
    capability.audit.retain_last = 1_001;
    let error = AuditRecorder::new(Some(&store))
        .record(&capability, &event(true))
        .unwrap_err();
    assert_eq!(error.to_string(), "record capability audit: internal");
}

pub(super) fn assert_cap_055_through_062_audit_contract() {
    recorder_skips_absent_store_and_disabled_policy();
    recorder_persists_only_sanitized_bounded_shape();
    recorder_reports_sanitized_storage_failure();
}

fn event(success: bool) -> AuditEvent {
    AuditEvent {
        operation: "SECRET method".to_owned(),
        target: "https://example.com/private?token=SECRET".to_owned(),
        success,
        error_class: "SECRET class".to_owned(),
        duration_millis: i64::MAX,
        request_bytes: -1,
        response_bytes: i64::MAX,
        redirects: i64::MAX,
    }
}
